sphinx/builders/linkcheck.py

```diff
@@ -313,71 +307,97 @@ def check_uri() -> tuple[str, str, int]:
+                    **kwargs,
+                )
+
+            # Linkcheck HTTP request logic:
+            #
+            # - Attempt HTTP HEAD before HTTP GET unless page content is required.
+            # - Follow server-issued HTTP redirects.
+            # - Respect server-issued HTTP 429 backoffs.
+            error_message = None
+            for retrieval_attempt in retrieval_methods():
+                try:
+                    with retrieval_attempt() as response:
+                        if response.ok and anchor and not contains_anchor(response, anchor):
+                            raise Exception(__(f'Anchor {anchor!r} not found'))
+
+                    # Copy data we need from the (closed) response
+                    status_code, redirect_status_code, retry_after, res_url = (
+                        response.status_code,
+                        response.history[-1].status_code if response.history else None,
+                        response.headers.get('Retry-After'),
+                        response.url,
+                    )
+                    response.raise_for_status()
+                    break
+
+                except (ConnectionError, TooManyRedirects) as err:
                     # Servers drop the connection on HEAD requests, causing
                     # ConnectionError.
-                    except (ConnectionError, HTTPError, TooManyRedirects) as err:
-                        if isinstance(err, HTTPError) and err.response.status_code == 429:
-                            raise
-                        # retry with GET request if that fails, some servers
-                        # don't like HEAD requests.
-                        with requests.get(req_url, stream=True, config=self.config,
-                                          auth=auth_info, **kwargs) as response:
-                            response.raise_for_status()
-            except HTTPError as err:
-                if err.response.status_code == 401:
-                    # We'll take "Unauthorized" as working.
-                    return 'working', ' - unauthorized', 0
-                elif err.response.status_code == 429:
-                    next_check = self.limit_rate(err.response)
-                    if next_check is not None:
-                        self.wqueue.put(CheckRequest(next_check, hyperlink), False)
-                        return 'rate-limited', '', 0
-                    return 'broken', str(err), 0
-                elif err.response.status_code == 503:
-                    # We'll take "Service Unavailable" as ignored.
-                    return 'ignored', str(err), 0
-                else:
+                    error_message = str(err)
+                    continue
+
+                except HTTPError as err:
+                    error_message = str(err)
+
+                    # Unauthorized: the reference probably exists
+                    if status_code == 401:
+                        return 'working', 'unauthorized', 0
```

Comment:
I'm not completely convinced that this logic is sound.  It was proposed in https://github.com/sphinx-doc/sphinx/issues/1687 - and for many webservers, yes perhaps HTTP 401 could indicate that a resource exists but is unavailable due to a lack of authentication - but my expectation would be that a decently large number of webservers apply that kind of response code across entire URL path ranges, in a way that does not confirm to the client whether a specific individual URL within that path range exists.

Reply:
Filed #11433 to discuss and track this.